python_toolbox/interpolate_log_file.py

```diff
@@ -0,0 +1,179 @@
+	if ~(sy < 1e-6):
+		pose_vec[0] = math.atan2(r21, r22)
+		pose_vec[1] = math.atan2(-r20, sy)
+		pose_vec[2] = math.atan2(r10, r00)
+	else:
+		pose_vec[0] = math.atan2(-r12, r11)
+		pose_vec[1] = math.atan2(-r20, sy)
+		pose_vec[2] = 0
+	pose_vec[3:] = pose[0:3, 3]
+	return pose_vec;
+
+def euler_2_rotation_matrix(x, y, z):
+	cosx = math.cos(x)
+	sinx = math.sin(x)
+	Rx = np.array([[1, 0, 0], [0, cosx, -sinx], [0, sinx, cosx]])
+	cosy = math.cos(y)
+	siny = math.sin(y)
+	Ry = np.array([[cosy, 0, siny], [0, 1, 0], [-siny, 0, cosy]])
+	cosz = math.cos(z)
+	sinz = math.sin(z)
+	Rz = np.array([[cosz, -sinz, 0], [sinz, cosz, 0], [0, 0, 1]])
+	return Rz.dot(Ry.dot(Rx))
+
+def transform_vector_6d_to_matrix_4d(pose_vec):
+	pose = np.identity(4)
+	pose[0:3, 3] = pose_vec[3:]
+	pose[0:3, 0:3] = euler_2_rotation_matrix(
+			pose_vec[0], pose_vec[1], pose_vec[2])
+	return pose
+
+if __name__ == "__main__":
+
+	print('')
+	print('==================================================================')
+	print('Python script for interpolating camera poses')
+	print('==================================================================')
+	print('Algorithm : ')
+	print('   1) Transform n-SE(3) camera pose matrices to nx6 matrix,')
+	print('      where each row contains euler angles and translation')
+	print('   2) Independently interpolate each column of nx6 matrix')
+	print('      using 1D cubic interpolation')
+	print('==================================================================')
+
+	if len(sys.argv) != 4:
+		print('Usage : python %s [input_log] [mapping_txt_file] [output_log]'
+				% sys.argv[0])
+		print('Example : python %s ../test_data/test.log ../test_data/mapping.txt ../test_data/test_interpolated.log'
+				% sys.argv[0])
+		print('')
+		print('Convention of [input_log]')
+		print('[frame ID] [frame ID] 0')
+		print('[R t]')
+		print('[0 1]')
+		print(': (repeats)')
+		print('')
+		print('Convention of [mapping_txt_file]')
+		print('[number of input camera poses]')
+		print('[number of desired number of interpolated poses]')
+		print('[Image ID] [video frame ID]')
+		print(': (repeats)')
+		quit()
+
+	# read files
+	trajectory = read_trajectory(sys.argv[1])
+	n_sampled_frames, n_total_frames, mapping = read_mapping(sys.argv[2])
+	print('%d camera poses are loaded' % n_sampled_frames)
+	print('Input poses are interpolated to %d poses' % n_total_frames)
+
+	# make nx6 matrix
+	n_trajectory = len(trajectory)
+	pose_matrix = np.zeros(shape = (n_trajectory, 6))
+	for iter in range(n_trajectory):
+		pose_vector = transform_matrix_4d_to_vector_6d(trajectory[iter].pose)
+		pose_matrix[iter,:] = pose_vector
+
+	# interpolation
+	pose_frame_desired = np.linspace(1, n_total_frames, n_total_frames)
+	pose_matrix_interpolation = np.zeros(shape = (n_total_frames, 6))
+	for iter in range(6):
+		pose_element_slice = pose_matrix[:, iter]
+		pose_frame_id = mapping[:, 1]
+		f = interp1d(pose_frame_id, pose_element_slice,
+				kind='cubic', fill_value='extrapolate')
```

Comment:
1. cubic interpolation and fill_value='extrapolate' doesn't work for me in f = interp1d(pose_frame_id, pose_element_slice, kind='cubic', fill_value='extrapolate'), works fine with kind='linear' though, linear should be OK, I think?
2. If we use only linear interpolation, we could use: < pose_matrix_interpolation[:, iter] = np.interp(pose_frame_desired, pose_frame_id, pose_element_slice) >, then users wouldn't have to install the scipy module?